### PR TITLE
Florian/sort todo

### PR DESCRIPTION
**Pull request by FloVanGH**
Add a header to the `todo` example that contains a `CheckBox` to sort the entries by name and a second `CheckBox` to filter the done items. Also implemented `set_row_data` for the `SortModel` and `FilterModel` (rust and cpp), to allow item data changes using that model types.

### CODE REVIEW

File: examples/todo/cpp/main.cpp [hunk 1]
```diff
@@ -44,5 +44,26 @@ int main()
         return slint::CloseRequestResponse::HideWindow;
     });
 
+    demo->set_is_header_visible(true);
+
+    demo->on_sort_filter([todo_model, demo = slint::ComponentWeakHandle(demo)] {
+        (*demo.lock())->set_todo_model(todo_model);
```
Comment (ogoffart) [inline examples/todo/cpp/main.cpp:50]: maybe you'd want to store demo.lock() in a local variable to make the code a bit more concise here

**Comment by ogoffart** [comment]
Also, please rebase this branch on origin/master so it only show the relevant commits.

File: examples/todo/ui/todo.slint [hunk 2]
```diff
@@ -69,6 +76,27 @@ MainWindow := Window {
...
 
+        HorizontalBox { 
+            visible: is-header-visible;
+            alignment: start; 
+
+            cb-sort-by-name := CheckBox {  
+                text: "Sort by name";
+
+                toggled => {
+                    sort_filter();
```
[the remark this answers is not in the document]
Comment (FloVanGH) [reply examples/todo/ui/todo.slint:87]: @ogoffart my intention was to combine both sort and filter. It works well together.

[the remark this answers is not in the document]
Comment (ogoffart) [reply examples/todo/ui/todo.slint:87]: Ah true, ignore my comment

**Review by ogoffart — COMMENTED** [review]
Somehow this branch contains still many duplicated commit with master.

You should do 
```sh
git fetch origin
git rebase origin/master
```


you can check with `gitk --all`  that it should look alright.

then you can `git push origin florian/sort-todo --force`  to update this branch.

File: examples/todo/ui/todo.slint [hunk 3]
```diff
@@ -69,6 +76,28 @@ MainWindow := Window {
...
+                }
+            }
+
+            CheckBox {  
+                text: "Hide done items";
+                checked <=> is-filter-done;
+
+                toggled => {
+                    sort_filter();
```
Comment (tronical) [inline examples/todo/ui/todo.slint:96]: The name of the callbacks might make innocent readers think "ah, this sorts the filter - I wonder what that means". I suggest to rename the callback to "apply_sorting_and_filtering()".

File: examples/todo/ui/todo.slint [hunk 4]
```diff
@@ -50,6 +53,10 @@ MainWindow := Window {
         { title: "Profit", checked: false },
     ];
 
+    property <bool> is-header-visible: false;
+    property <bool> is-sort-by-name: false;
+    property <bool> is-filter-done: false;
```
Comment (tronical) [inline examples/todo/ui/todo.slint:58]: This reads as if the property indicates if the "filter" is finished (done). I suggest to rename this to "filter-done-items" or "hide-done-items" (like the label).

File: examples/todo/ui/todo.slint [hunk 5]
```diff
@@ -50,6 +53,10 @@ MainWindow := Window {
         { title: "Profit", checked: false },
     ];
 
+    property <bool> is-header-visible: false;
```
Comment (tronical) [inline examples/todo/ui/todo.slint:56]: Unlike "sort-by-name" this doesn't suffer from verb ambiguity - I suggest to rename this to simply "header-visible" or "show-header".

**Comment by FloVanGH** [comment]
@ogoffart still contains the duplicated commits. I created this branch from the sortmodel branch before the sortmodel was squashed and merged in the master. Maybe this is related to the mixed history.

**Review by tronical — APPROVED** [review]
Please squash when merging.

**Comment by tronical** [comment]
> Implement `set_row_data` for `SortModel` and `FilterModel` for `Rust` and `cpp`. Add sort and filter header to the `todo` example.

This shows up in Tobias' weekly summary, which looks good. But the subject of this merge request won't look good in the weekly summary. I suggest to use the "Edit" button to change it to something a slightly catchier headline :)

**Comment by tronical** [comment]
I can do the rebase & merge if you're a bit unsure how to clean this up - just let me know :)

**Comment by FloVanGH** [comment]
@tronical sure you can help me with this. Just let me know what you have done to clean it again ;-)

File: internal/core/model/adapters.rs [hunk 6]
```diff
@@ -604,7 +609,7 @@ where
 ///     SharedString::from("dolor"),
 /// ]);
 ///
-/// let sorted_model = SortModel::new(model, |lhs, rhs| lhs.to_lowercase().cmp(&rhs.to_lowercase()));
+/// let sorted_model = SortModel::new(model, ≈));
```
Comment (tronical) [inline internal/core/model/adapters.rs:612]: This line looks unintentional:)

Comment (FloVanGH) [reply internal/core/model/adapters.rs]: Done thx

**Review by tronical — APPROVED** [review]
History looks clean, only todo stuff in it. Please squash when merging.

**Pull request merged**